crates/ide-assists/src/handlers/extract_variable.rs

```diff
@@ -58,9 +58,30 @@ pub(crate) fn extract_variable(acc: &mut Assists, ctx: &AssistContext<'_>) -> Op
     }
 
     let parent = to_extract.syntax().parent().and_then(ast::Expr::cast);
-    let needs_adjust = parent
-        .as_ref()
-        .map_or(false, |it| matches!(it, ast::Expr::FieldExpr(_) | ast::Expr::MethodCallExpr(_)));
+    // Any expression that autoderefs may need adjustment.
+    let mut needs_adjust = parent.as_ref().map_or(false, |it| match it {
```

Comment:
I wonder, couldn't we check if the type of `let ty = ctx.sema.type_of_expr(&to_extract).map(TypeInfo::adjusted);` and `let ty = ctx.sema.type_of_expr(&to_extract).map(TypeInfo::original);` differs to tell this? I guess we are only interested in autoderef adjustments so probably not